### PR TITLE
Separated user and admin functions in list methods of some entities.

### PR DESCRIPTION
**Pull request by hetjagani**

### CODE REVIEW

File: action/cart/list.go [hunk 1]
```diff
@@ -47,3 +48,48 @@ func list(w http.ResponseWriter, r *http.Request) {
...
+
+	tx := model.DB.Model(&model.CartItem{}).Preload("Product").Preload("Product.Currency").Preload("Product.FeaturedMedium").Preload("Product.Tags").Preload("Product.Datasets")
+
+	if userID != 0 {
+		tx.Where(&model.CartItem{
+			UserID: uint(userID),
+		}).Count(&result.Total).Offset(offset).Limit(limit).Find(&result.Nodes)
+	} else {
+		tx.Count(&result.Total).Offset(offset).Limit(limit).Find(&result.Nodes)
+	}
```
Comment (dmonark) [inline action/cart/list.go]: keep if condition for adding userID in where operation and all other operation are common so keep it outside.

File: action/cart/list.go [hunk 2]
```diff
@@ -28,7 +29,7 @@ type paging struct {
 // @Param page query string false "page number"
 // @Success 200 {object} paging
 // @Router /cartitems [get]
-func list(w http.ResponseWriter, r *http.Request) {
+func userlist(w http.ResponseWriter, r *http.Request) {
```
Comment (dmonark) [inline action/cart/list.go]: change the function name to mylist

File: action/cart/list.go [hunk 3]
```diff
@@ -47,3 +48,48 @@ func list(w http.ResponseWriter, r *http.Request) {
...
+// @Tags Cart
+// @ID get-all-carts
+// @Produce  json
+// @Param X-User header string false "User ID"
+// @Param user query string false "User ID"
+// @Param limit query string false "limt per page"
+// @Param page query string false "page number"
+// @Success 200 {object} paging
+// @Router /cartitems [get]
+func adminlist(w http.ResponseWriter, r *http.Request) {
```
Comment (dmonark) [inline action/cart/list.go]: change the function name to `list`

File: action/membership/list.go [hunk 4]
```diff
@@ -14,24 +18,79 @@ type paging struct {
...
 // @Tags Membership
 // @ID get-all-memberships
 // @Produce  json
+// @Param X-User header string true "User ID"
 // @Param limit query string false "limt per page"
 // @Param page query string false "page number"
 // @Success 200 {object} paging
 // @Router /memberships [get]
-func list(w http.ResponseWriter, r *http.Request) {
+func userlist(w http.ResponseWriter, r *http.Request) {
```
Comment (dmonark) [inline action/membership/list.go]: change it to mylist

File: action/membership/list.go [hunk 5]
```diff
@@ -14,24 +18,79 @@ type paging struct {
...
+// @Tags Membership
+// @ID get-all-memberships
+// @Produce  json
+// @Param X-User header string true "User ID"
+// @Param user query string false "User ID"
+// @Param limit query string false "limt per page"
+// @Param page query string false "page number"
+// @Success 200 {object} paging
+// @Router /memberships [get]
+func adminlist(w http.ResponseWriter, r *http.Request) {
```
Comment (dmonark) [inline action/membership/list.go]: change it to list

File: action/order/list.go [hunk 6]
```diff
@@ -14,24 +18,79 @@ type paging struct {
...
 // @Tags Order
 // @ID get-all-orders
 // @Produce  json
+// @Param X-User header string true "User ID"
 // @Param limit query string false "limt per page"
 // @Param page query string false "page number"
 // @Success 200 {object} paging
 // @Router /orders [get]
-func list(w http.ResponseWriter, r *http.Request) {
+func userlist(w http.ResponseWriter, r *http.Request) {
```
Comment (dmonark) [inline action/order/list.go]: change it to mylist

File: action/order/list.go [hunk 7]
```diff
@@ -14,24 +18,79 @@ type paging struct {
...
+// @Tags Order
+// @ID get-all-orders
+// @Produce  json
+// @Param X-User header string true "User ID"
+// @Param user query string false "User ID"
+// @Param limit query string false "limt per page"
+// @Param page query string false "page number"
+// @Success 200 {object} paging
+// @Router /orders [get]
+func adminlist(w http.ResponseWriter, r *http.Request) {
```
Comment (dmonark) [inline action/order/list.go]: change it to list

File: action/membership/list.go [hunk 8]
```diff
@@ -14,24 +18,79 @@ type paging struct {
...
+
+	tx := model.DB.Model(&model.Membership{}).Preload("Plan").Preload("Plan.Catalogs").Preload("Payment").Preload("Payment.Currency")
+
+	if userID != 0 {
+		tx.Where(&model.Membership{
+			UserID: uint(userID),
+		}).Count(&result.Total).Offset(offset).Limit(limit).Find(&result.Nodes)
+	} else {
+		tx.Count(&result.Total).Offset(offset).Limit(limit).Find(&result.Nodes)
+	}
```
Comment (dmonark) [inline action/membership/list.go:93]: keep if condition for adding userID in where operation and all other operation are common so keep it outside.

File: action/order/list.go [hunk 9]
```diff
@@ -14,24 +18,79 @@ type paging struct {
...
+
+	tx := model.DB.Preload("Payment").Preload("Payment.Currency").Preload("Products").Preload("Products.Datasets").Preload("Products.Tags").Model(&model.Order{})
+
+	if userID != 0 {
+		tx.Where(&model.Order{
+			UserID: uint(userID),
+		}).Count(&result.Total).Offset(offset).Limit(limit).Find(&result.Nodes)
+	} else {
+		tx.Count(&result.Total).Offset(offset).Limit(limit).Find(&result.Nodes)
+	}
```
Comment (dmonark) [inline action/order/list.go:93]: keep if condition for adding userID in where operation and all other operation are common so keep it outside.

**Review by dmonark — CHANGES_REQUESTED** [review]

File: action/membership/create.go [hunk 10]
```diff
@@ -56,6 +56,17 @@ func create(w http.ResponseWriter, r *http.Request) {
...
+	// Check if the plan is not deleted
+	plan := model.Plan{}
+	plan.ID = membership.PlanID
+	if err = tx.Model(&model.Plan{}).First(&plan).Error; err != nil {
+		tx.Rollback()
+		loggerx.Error(err)
+		errorx.Render(w, errorx.Parser(errorx.CannotSaveChanges()))
+		return
+	}
+
```
Comment (shreeharsha-factly) [inline action/membership/create.go:69]: As plan id is a Foreign Key if it does not exists or deleted then Membership cannot be created. I don't think this query is required.

File: action/cart/list.go (hunk 1 above)
Comment (hetjagani) [reply action/cart/list.go]: We need limit and offset after count.. Similar problem was faced by Harsha in dega-server post/list.go, so at that time to resolve the bug we had made changes like this.

File: action/membership/create.go (hunk 10 above)
Comment (hetjagani) [reply action/membership/create.go:69]: But when we delete a plan, it does not actually delete the plan, we are doing soft delete so the deleted_at field is set. Hence, the foreign key constrain does not fail when we add some plan_id which is deleted previously.

Comment (shreeharsha-factly) [reply action/membership/create.go:69]: Are you sure about that? If not once check it and make changes accordingly.

Comment (hetjagani) [reply action/membership/create.go:69]: Yaa I checked it.... I came across this bug while testing only so I fixed it.

**Review by shreeharsha-factly — APPROVED** [review]

**Review by dmonark — APPROVED** [review]

**Pull request merged**